# ``Statistics Exercise`` - **``Solution``**

## __Import Library__

In [3]:
import pandas as pd
pd.set_option('display.max_columns', None)

## __Open Dataset__

In [4]:
df = pd.read_csv('melb_data.csv')
df.head()

,Suburb,Address,Rooms,Type,Price,Method,SellerG,Date,Distance,Postcode,Bedroom2,Bathroom,Car,Landsize,BuildingArea,YearBuilt,CouncilArea,Lattitude,Longtitude,Regionname,Propertycount
0,Abbotsford,85 Turner St,2,h,1480000.0,S,Biggin,3/12/2016,2.5,3067.0,2.0,1.0,1.0,202.0,NaN,NaN,Yarra,-37.7996,144.9984,Northern Metropolitan,4019.0
1,Abbotsford,25 Bloomburg St,2,h,1035000.0,S,Biggin,4/02/2016,2.5,3067.0,2.0,1.0,0.0,156.0,79.0,1900.0,Yarra,-37.8079,144.9934,Northern Metropolitan,4019.0
2,Abbotsford,5 Charles St,3,h,1465000.0,SP,Biggin,4/03/2017,2.5,3067.0,3.0,2.0,0.0,134.0,150.0,1900.0,Yarra,-37.8093,144.9944,Northern Metropolitan,4019.0
3,Abbotsford,40 Federation La,3,h,850000.0,PI,Biggin,4/03/2017,2.5,3067.0,3.0,2.0,1.0,94.0,NaN,NaN,Yarra,-37.7969,144.9969,Northern Metropolitan,4019.0
4,Abbotsford,55a Park St,4,h,1600000.0,VB,Nelson,4/06/2016,2.5,3067.0,3.0,1.0,2.0,120.0,142.0,2014.0,Yarra,-37.8072,144.9941,Northern Metropolitan,4019.0


<hr>

## **Nomor 1**

Tampilkan statistik deskriptif pada variabel **Rooms, Distance, Landsize**, dan **Price**!

In [3]:
df[['Rooms','Distance', 'Landsize', 'Price']].describe()

,Rooms,Distance,Landsize,Price
count,13580.000000,13580.000000,13580.000000,1.358000e+04
mean,2.937997,10.137776,558.416127,1.075684e+06
std,0.955748,5.868725,3990.669241,6.393107e+05
min,1.000000,0.000000,0.000000,8.500000e+04
25%,2.000000,6.100000,177.000000,6.500000e+05
50%,3.000000,9.200000,440.000000,9.030000e+05
75%,3.000000,13.000000,651.000000,1.330000e+06
max,10.000000,48.100000,433014.000000,9.000000e+06


,Rooms,Distance,Landsize,Price
count,13580.000000,13580.000000,13580.000000,1.358000e+04
mean,2.937997,10.137776,558.416127,1.075684e+06
std,0.955748,5.868725,3990.669241,6.393107e+05
min,1.000000,0.000000,0.000000,8.500000e+04
25%,2.000000,6.100000,177.000000,6.500000e+05
50%,3.000000,9.200000,440.000000,9.030000e+05
75%,3.000000,13.000000,651.000000,1.330000e+06
max,10.000000,48.100000,433014.000000,9.000000e+06


<hr>

## **Nomor 2**

a) Berapa jumlah **Rooms** pada rumah yang harganya paling mahal?


In [5]:
df[df['Price'] == df['Price'].max()][['Rooms']]

,Rooms
12094,3


,Rooms
12094,3


b) Berapa rata-rata harga rumah yang jarak ke CBD (variabel **Distance**) terdekat?


In [7]:
df[df['Distance'] ==  df['Distance'].min()][['Price']].mean().item()

671166.6666666666

671166.6666666666

c) Berapa rata-rata harga rumah yang luas tanahnya (variabel **Landsize**) 1 kali standar deviasi di atas rata-rata luas tanah?


In [6]:
df[df['Landsize'] >= df['Landsize'].mean() + df['Landsize'].std()]['Price'].mean().item()

880776.6666666666

880776.6666666666

d) Berapa rentang (_range_) harga rumah (variabel **Price**) pada region ``Southern Metropolitan ``?

In [8]:
df[df['Regionname'] == 'Southern Metropolitan']['Price'].max() - df[df['Regionname'] == 'Southern Metropolitan']['Price'].min()

7869000.0

7869000.0

<hr>

## **Nomor 3**

* Nilai outlier adalah nilai yang berada di bawah ``Q1 - (1.5 * IQR)`` atau di atas ``Q3 + (1.5 * IQR)``
* Hitung dulu Q1, Q2, dan IQR pada pada kolom **Price** (harga rumah)
* Hitung ada berapa nilai outlier pada kolom **Price**!

In [14]:
import numpy as np

df['Price'] = df['Price'].apply(np.int64)

In [15]:
df['Price'] = df['Price'].apply(np.int64)
Q1 = df['Price'].quantile(0.25)
Q3 = df['Price'].quantile(0.75)
IQR = Q3 - Q1

low = Q1 - (1.5 * IQR) 
up = Q3 + (1.5 * IQR)

outL = df[(df['Price'] < low) | (df['Price'] > up)]

len(outL)

612

In [23]:
IQR.item()

680000.0

680000.0

In [20]:
outL['Price'].head()

80     2850000
92     2615000
97     2575000
102    3010000
103    2800000
Name: Price, dtype: int64

,Price
80,2850000
92,2615000
97,2575000
102,3010000
103,2800000


In [ ]:
# jumlah outlier pada kolom 'Price'


612

In [25]:
# lokasi
outL.index
# untuk mengetahui posisi outlier

Index([   80,    92,    97,   102,   103,   108,   112,   118,   124,   233,
       ...
       13466, 13468, 13474, 13483, 13518, 13521, 13523, 13553, 13555, 13578],
      dtype='int64', length=612)

<hr>

## **Nomor 4**

#### **``Shapiro Wilk Test``**

Menguji asumsi normalitas variabel ``Price``

#### **Interpretation**
* If the P-Value of the Shapiro Wilk Test is larger than 0.05, we assume a normal distribution
* If the P-Value of the Shapiro Wilk Test is smaller than 0.05, we do not assume a normal distribution

In [26]:
from scipy import stats

# Fungsi mengembalikan (Test Statistic, P-value)
stat, p_value = stats.shapiro(df['Price'])


if p_value > 0.05:
    print(f"P-value: {p_value}. So, We assume a normal distribution.")
    
else:
    print(f"P-value: {p_value}. So, We do not assume a normal distribution.")

P-value: 1.4465718615128594e-79. So, We do not assume a normal distribution.


/opt/anaconda3/lib/python3.13/site-packages/scipy/stats/_axis_nan_policy.py:586: UserWarning: scipy.stats.shapiro: For N > 5000, computed p-value may not be accurate. Current N is 13580.
  res = hypotest_fun_out(*samples, **kwds)


P-value: 0.0. So, we do not assume a normal distribution


c:\users\moham\appdata\local\programs\python\python38\lib\site-packages\scipy\stats\morestats.py:1681: UserWarning: p-value may not be accurate for N > 5000.
  warnings.warn("p-value may not be accurate for N > 5000.")


#### **``D’Agostino and Pearson’s Test``**

Menguji asumsi normalitas variabel ``Price``

#### **Interpretation**
* If the P-Value of the D'Agostino-Pearson Test is larger than 0.05, we assume a normal distribution
* If the P-Value of the D'Agostino-Pearson Test is smaller than 0.05, we do not assume a normal distribution

In [27]:
from scipy import stats

# Fungsi mengembalikan (Test Statistic, P-value)
stat, p_value = stats.normaltest(df['Price'])

if p_value > 0.05:
    print(f"P-value: {p_value}. So, We assume a normal distribution.")
    
else:
    print(f"P-value: {p_value}. So, We do not assume a normal distribution.")

P-value: 0.0. So, We do not assume a normal distribution.


P-value: 0.0. So, we do not assume a normal distribution


<hr>

## **Nomor 5**

Jika diasumsikan telah memenuhi syarat statistik parametrik, hitung seberapa kuat korelasi variabel **Rooms** dengan **Price**, serta arah korelasinya!

In [28]:
df['Rooms'].corr(df['Price']).item()

0.496633676186531

0.49663367618652987

<hr>

## **Nomor 6**

Jika diasumsikan telah memenuhi syarat statistik parametrik, dari beberapa Regionname manakah **Regionname** yang variabel jumlah kamarnya (**Rooms**) memiliki korelasi terkuat dengan variabel harga rumah (**Price**)!

In [32]:
result = []

regions = df['Regionname'].unique()

for region in regions:
    temp = df[df['Regionname'] == region]

    # hitung korelasi Rooms & Price
    corr = temp['Rooms'].corr(temp['Price'])

    # jika korelasi bernilai angka valid (bukan NaN)
    if not np.isnan(corr):
        result.append(region)

df['Regionname'].unique()


array(['Northern Metropolitan', 'Western Metropolitan',
       'Southern Metropolitan', 'Eastern Metropolitan',
       'South-Eastern Metropolitan', 'Eastern Victoria',
       'Northern Victoria', 'Western Victoria'], dtype=object)

array(['Northern Metropolitan', 'Western Metropolitan',
       'Southern Metropolitan', 'Eastern Metropolitan',
       'South-Eastern Metropolitan', 'Eastern Victoria',
       'Northern Victoria', 'Western Victoria'], dtype=object)

In [33]:
hasil = {}

for region in regions:
    data_region = df[df['Regionname'] == region]
    hasil[region] = data_region['Rooms'].corr(data_region['Price'])

for regionn,korelasii in hasil.items():
    print(regionn, "=", korelasii)

Northern Metropolitan = 0.39281345830224085
Western Metropolitan = 0.4760302938434976
Southern Metropolitan = 0.6837438462796436
Eastern Metropolitan = 0.5029928557468851
South-Eastern Metropolitan = 0.26779944081140455
Eastern Victoria = 0.31442339534695335
Northern Victoria = 0.3611442989081275
Western Victoria = 0.5518865952271457


Northern Metropolitan = 0.39281345830224074
Western Metropolitan = 0.47603029384349793
Southern Metropolitan = 0.6837438462796437
Eastern Metropolitan = 0.5029928557468845
South-Eastern Metropolitan = 0.2677994408114044
Eastern Victoria = 0.31442339534695335
Northern Victoria = 0.36114429890812727
Western Victoria = 0.5518865952271457


In [34]:
corr_df = (
    pd.DataFrame(list(hasil.items()), columns=['', 'Correlation Price vs Rooms'])
    .sort_values(by='Correlation Price vs Rooms', ascending=False)
    .reset_index(drop=True) # untuk memulai dari 0
)

corr_df

,,Correlation Price vs Rooms
0,Southern Metropolitan,0.683744
1,Western Victoria,0.551887
2,Eastern Metropolitan,0.502993
3,Western Metropolitan,0.476030
4,Northern Metropolitan,0.392813
5,Northern Victoria,0.361144
6,Eastern Victoria,0.314423
7,South-Eastern Metropolitan,0.267799


,Correlation Price vs Rooms
Southern Metropolitan,0.683744
Western Victoria,0.551887
Eastern Metropolitan,0.502993
Western Metropolitan,0.476030
Northern Metropolitan,0.392813
Northern Victoria,0.361144
Eastern Victoria,0.314423
South-Eastern Metropolitan,0.267799


<hr>

## **Nomor 7**

Jika diasumsikan **tidak** memenuhi syarat statistik parametrik, hitung seberapa kuat korelasi variabel **Distance** dengan **Price**, serta arah korelasinya!

In [17]:
corr = df['Distance'].corr(df['Price'], method='spearman')
corr.item()

-0.12998984131880323

-0.12998984131880323

<hr>

## **Nomor 8**

Jika diasumsikan telah memenuhi syarat statistik parametrik, dari beberapa Regionname manakah **Regionname** yang variabel jumlah jarak ke CBD (**Distance**) paling tidak berkorelasi dengan variabel harga rumah (**Price**)! Serta tampilkan juga Regionname yang paling berkorelasi kedua variabel ini!

In [18]:
groups = df.groupby('Regionname')

corrs = {}

for name, g in groups:
    corrs[name] = g['Distance'].corr(g['Price'])

for regionn,korelasii in corrs.items():
    print(regionn, "=", korelasii)

Eastern Metropolitan = -0.12435562852683703
Eastern Victoria = -0.34569916122464583
Northern Metropolitan = -0.368551996179051
Northern Victoria = 0.08956649327360577
South-Eastern Metropolitan = -0.28556405189080497
Southern Metropolitan = 0.0024891492189334026
Western Metropolitan = -0.33408555047092137
Western Victoria = -0.04606415345276121


Northern Metropolitan = -0.368551996179051
Western Metropolitan = -0.33408555047092137
Southern Metropolitan = 0.002489149218933464
Eastern Metropolitan = -0.12435562852683693
South-Eastern Metropolitan = -0.2855640518908052
Eastern Victoria = -0.34569916122464606
Northern Victoria = 0.08956649327360579
Western Victoria = -0.04606415345276126


In [21]:
corr_df = (
    pd.DataFrame(list(corrs.items()), columns=['', 'Correlation Price vs Rooms'])
    .sort_values(by='Correlation Price vs Rooms', ascending=True)
    .reset_index(drop=True)
)

corr_df

,,Correlation Price vs Rooms
0,Northern Metropolitan,-0.368552
1,Eastern Victoria,-0.345699
2,Western Metropolitan,-0.334086
3,South-Eastern Metropolitan,-0.285564
4,Eastern Metropolitan,-0.124356
5,Western Victoria,-0.046064
6,Southern Metropolitan,0.002489
7,Northern Victoria,0.089566


,Correlation Price vs Distance
Northern Metropolitan,-0.368552
Eastern Victoria,-0.345699
Western Metropolitan,-0.334086
South-Eastern Metropolitan,-0.285564
Eastern Metropolitan,-0.124356
Western Victoria,-0.046064
Southern Metropolitan,0.002489
Northern Victoria,0.089566


* Regionname yang paling tidak berkorelasi atau korelasinya paling lemah di dua variabel di atas adalah **Sourthern Metropolitan** (0.002489 atau 0.24 persen).
* Regionname yang paling berkorelasi di dua variabel di atas adalah **Northern Metropolitan** (-0.368552 atau - 36,86 persen)

<hr>

## **Nomor 9**

Uji hipotesis apakah metode pembelian/**Method** berkaitan dengan **Type** / tipe rumah (dependen) atau tidak saling berkaitan (independen)!

In [36]:
table = pd.crosstab(df['Type'], df['Method'])
table

Method,PI,S,SA,SP,VB
Type,,,,,
h,1069,6507,66,1079,728
t,134,723,7,143,107
u,361,1792,19,481,364


Method,PI,S,SA,SP,VB
Type,,,,,
h,1069,6507,66,1079,728
t,134,723,7,143,107
u,361,1792,19,481,364


#### **Hypothesis:**
* Ho : The two variables are independent
* Ha :  The two variables are dependent.

In [37]:
# uji coba untuk hipotesis yang datanya kategorikal
from scipy.stats import chi2_contingency

chi2, p, dof, expected = chi2_contingency(table)

print(f"P value: {p}")

if p < 0.05:
    print("Dependent (tolak H0)")
    
else:
    print("Independent (terima H0)")

P value: 3.183702485154384e-22
Dependent (tolak H0)


P value: 3.183702485154384e-22
Dependent (tolak H0)


<hr>

## **Nomor 10**

Uji hipotesis apakah rata-rata harga rumah (**Price**) antar Regionname berbeda ataukah sama!

Diasumsikan persyaratan statistik parametrik terpenuhi

**Hipotesis**:
* Ho : μ1 = μ2 = … = μk (Rata-rata harga rumah antar Regionname sama)  
* Ha : Setidaknya ada satu pasang Regionname yang rata-rata harga rumahnya tidak sama, tapi kita tidak tahu yang mana

In [38]:
from scipy import stats

# 1. Cek semua Regionname unik
regions = df['Regionname'].unique()
print(regions)

['Northern Metropolitan' 'Western Metropolitan' 'Southern Metropolitan'
 'Eastern Metropolitan' 'South-Eastern Metropolitan' 'Eastern Victoria'
 'Northern Victoria' 'Western Victoria']


array(['Northern Metropolitan', 'Western Metropolitan',
       'Southern Metropolitan', 'Eastern Metropolitan',
       'South-Eastern Metropolitan', 'Eastern Victoria',
       'Northern Victoria', 'Western Victoria'], dtype=object)

In [39]:
price_groups = [df[df['Regionname'] == region]['Price'] for region in regions]

stats.f_oneway(*price_groups)

# anova untuk mencari perbandingan rata" antara beberapa kelompok (dalam konteks ini price antar region)

F_onewayResult(statistic=np.float64(284.40018651008097), pvalue=np.float64(0.0))

F_onewayResult(statistic=284.40018651008097, pvalue=0.0)

Jika p-value kurang dari sama dengan 0.05 maka cukup bukti untuk menolak Ho.
Berarti Ha diterima ada pasangan Regionname yang perbedaan rata-rata harga rumahnya signifikan, tapi tidak tahu yang mana.

In [26]:
price_mean = (
    df.groupby('Regionname')['Price'].mean()
    .reset_index()
    .rename(columns={'Price': 'Price Mean'})
    .sort_values(by='Price Mean', ascending=False)
    .reset_index(drop=True)
)
price_mean

,Regionname,Price Mean
0,Southern Metropolitan,1.372963e+06
1,Eastern Metropolitan,1.104080e+06
2,South-Eastern Metropolitan,9.229438e+05
3,Northern Metropolitan,8.981711e+05
4,Western Metropolitan,8.664205e+05
5,Eastern Victoria,6.999808e+05
6,Northern Victoria,5.948293e+05
7,Western Victoria,3.975234e+05


,Price Mean
Southern Metropolitan,1372963.37
Eastern Metropolitan,1104079.63
South-Eastern Metropolitan,922943.78
Northern Metropolitan,898171.08
Western Metropolitan,866420.52
Eastern Victoria,699980.79
Northern Victoria,594829.27
Western Victoria,397523.44


<hr>

## **Thank you & Keep Learning**

In [44]:
%%writefile test.txt
this is the first line

Writing test.txt


In [57]:
repeat_lyrics()
def repeat_lyrics(): 
    print_lyrics() 
    print_lyrics()
    
def print_lyrics():
    print("Cicak-cicak di dinding.")
    print( 'Diam-diam merayap.')

Cicak-cicak di dinding.
Diam-diam merayap.
Cicak-cicak di dinding.
Diam-diam merayap.
